libyul/backends/evm/StackLayoutGenerator.cpp

```diff
@@ -0,0 +1,567 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+/**
+ * Stack layout generator for Yul to EVM code generation.
+ */
+
+#include <libyul/backends/evm/StackLayoutGenerator.h>
+
+#include <libyul/backends/evm/OptimizedEVMCodeTransform.h>
+#include <libyul/backends/evm/StackHelpers.h>
+
+#include <libsolutil/Algorithms.h>
+#include <libsolutil/cxx20.h>
+#include <libsolutil/Permutations.h>
+#include <libsolutil/Visitor.h>
+
+#include <range/v3/algorithm/any_of.hpp>
+#include <range/v3/range/conversion.hpp>
+#include <range/v3/view/all.hpp>
+#include <range/v3/view/concat.hpp>
+#include <range/v3/view/drop.hpp>
+#include <range/v3/view/drop_last.hpp>
+#include <range/v3/view/filter.hpp>
+#include <range/v3/view/map.hpp>
+#include <range/v3/view/reverse.hpp>
+#include <range/v3/view/take.hpp>
+#include <range/v3/view/take_last.hpp>
+#include <range/v3/view/transform.hpp>
+
+using namespace solidity;
+using namespace solidity::yul;
+using namespace std;
+
+StackLayoutGenerator::StackLayoutGenerator(StackLayout& _layout): m_layout(_layout)
+{
+}
+
+namespace
+{
+struct PreviousSlot { size_t slot; };
+
+// TODO: Rewrite this as custom algorithm matching createStackLayout exactly and make it work
+// for all cases, including duplicates and removals of slots that can be generated on the fly, etc.
+// After that the util::permute* functions can be removed.
```

Comment:
I think this is among the most important ``TODO``s right now.

Reply:
For some more context: the situation is that we know what stack layout we need to have *after* an operation (call it "post-layout"), we know the input and output of that operation and we want to calculate the ideal stack layout we want *before* the operation, s.t. executing the operation allows us to generate the post-layout with minimal swapping in the ``createStackLayout`` algorithm (which in itself has a few TODOs and doesn't handle all cases optimally yet).

The idea of the current implementation is to start from a vector referring to the slots that are previously on the stack with an increasing sequence of integers, which are allowed to end up anywhere in the stack, and append the output of the operation to that. Since for all output slots of the operation, we know where they are supposed to end up, we then move them to their place (abusing the permutation algorithm that is not really designed for this) while allowing any position for the "previous slots". In the end we can see how the previous slots were permuted and invert that permutation to get the ideal stack layout.

Example:
Consider the operation ``x := add(a,b)``.
Assume that after the operation we want the stack layout: ``[b x c]``.
So we know we must keep ``b`` and ``c`` on stack additionally to the arguments of ``add`` themselves - but we don't know in which order.
So the previous stack layout is:
``[?1 ?2 b a]``
i.e. some unknown slots for which we track the order plus the input arguments of the operation.
Now we assume the operation is executed and get
``[?1 ?2 x]``
From this we want to get to ``[b x c]``.
So we want to swap ``x`` into place, yielding:
``[?1 x ?2]``
Which we can compare to ``[b x c]``, meaning that ``?1`` has to be ``b`` and ``?2`` has to be ``c``, so our ideal start layout is:
``[b c a b]``.


This concept, I think, is good, but the implementation so far sucks. The shuffling from ``[?1 ?2 x]`` to the resulting layout should be able to account for duplicated slots in the target and for slots that can be freely generated in the target and be completely in sync with ``createStackLayout``.

Reply:
Yeah, restricted to swapping only this is easy - but I think we don't get around having a version that can take dupping into account - otherwise we'll always get annoying suboptimal cases (depending on the case dupping at the wrong time can be *really* bad)...